File: school-ms/frontend/src/services/consolidatedCourseService.ts
import { api } from './api';

/**
 * Consolidated Course interface that matches the backend ConsolidatedCourse model
 */
export interface ConsolidatedCourse {
  id?: number;
  name: string;
  department: string;
  teacherId: number;
  teacherName?: string; // Transient field for display
  credits: number;
  capacity: number;
  enrolled: number;
  createdAt?: string;
  updatedAt?: string;
}

/**
 * Consolidated Course Service that works with the new consolidated backend endpoints.
 * This service should replace the existing courseService over time.
 */
export const consolidatedCourseService = {
  // Get all courses
  getAll: () => api.get<ConsolidatedCourse[]>('courses'),
  
  // Get course by ID
  getById: (id: number) => api.get<ConsolidatedCourse>(`courses/${id}`),
  
  // Get courses by department
  getByDepartment: (department: string) => api.get<ConsolidatedCourse[]>(`courses/department/${department}`),
  
  // Get courses by teacher ID
  getByTeacherId: (teacherId: number) => api.get<ConsolidatedCourse[]>(`courses/teacher/${teacherId}`),
  
  // Create a new course
  create: (course: ConsolidatedCourse) => api.post<ConsolidatedCourse>('courses', course),
  
  // Update an existing course
  update: (id: number, course: ConsolidatedCourse) => api.put<ConsolidatedCourse>(`courses/${id}`, course),
  
  // Delete a course
  delete: (id: number) => api.delete(`courses/${id}`),
  
  // Get course statistics
  getStats: () => api.get<{
    totalCourses: number,
    totalCapacity: number,
    totalEnrolled: number,
    fillRate: string
  }>('courses/stats')
};

export default consolidatedCourseService;

File: school-ms/backend/school-app/src/main/java/com/school/course/service/ConsolidatedCourseService.java
package com.school.course.service;

import com.school.course.model.ConsolidatedCourse;
import java.util.List;

/**
 * Consolidated Course service that combines functionality from both previous implementations.
 * This combines features from:
 * - com.schoolms.service.CourseService
 * - com.school.course.service.CourseService
 */
public interface ConsolidatedCourseService {
    /**
     * Get all courses in the system
     * 
     * @return List of all courses
     */
    List<ConsolidatedCourse> getAllCourses();

    /**
     * Get course by ID
     * 
     * @param id Course ID
     * @return Course if found, throws exception if not found
     */
    ConsolidatedCourse getCourseById(Long id);

    /**
     * Create a new course
     * 
     * @param course Course to create
     * @return Created course with ID
     */
    ConsolidatedCourse createCourse(ConsolidatedCourse course);

    /**
     * Update an existing course
     * 
     * @param id Course ID to update
     * @param course Updated course information
     * @return Updated course
     */
    ConsolidatedCourse updateCourse(Long id, ConsolidatedCourse course);

    /**
     * Delete a course
     * 
     * @param id Course ID to delete
     */
    void deleteCourse(Long id);

    /**
     * Get courses by department
     * 
     * @param department Department name
     * @return List of courses in the department
     */
    List<ConsolidatedCourse> getCoursesByDepartment(String department);

    /**
     * Get courses by teacher
     * 
     * @param teacherId Teacher ID
     * @return List of courses taught by the teacher
     */
    List<ConsolidatedCourse> getCoursesByTeacherId(Long teacherId);
}
